cnd/src/swap_protocols/rfc003/alice/mod.rs

```diff
@@ -52,56 +45,50 @@ pub enum SwapCommunication<AL: Ledger, BL: Ledger, AA: Asset, BA: Asset> {
-        bob_dial_info: DialInformation,
-        save_state: Arc<dyn SaveState<AL, BL, AA, BA>>,
-    ) -> Box<FutureSwapOutcome<AL, BL, AA, BA>> {
-        let swap_request = self.request();
-        let start_state = Start {
-            id: swap_request.id,
-            alpha_ledger: swap_request.alpha_ledger,
-            beta_ledger: swap_request.beta_ledger,
-            alpha_asset: swap_request.alpha_asset,
-            beta_asset: swap_request.beta_asset,
-            hash_function: swap_request.hash_function,
-            alpha_ledger_refund_identity: swap_request.alpha_ledger_refund_identity,
-            beta_ledger_redeem_identity: swap_request.beta_ledger_redeem_identity,
-            alpha_expiry: swap_request.alpha_expiry,
-            beta_expiry: swap_request.beta_expiry,
-            secret_hash: swap_request.secret_hash,
-        };
-
-        let context = Context {
-            alpha_ledger_events,
-            beta_ledger_events,
-            communication_events: Box::new(AliceToBob::new(comit_client, bob_dial_info)),
-            state_repo: save_state,
-        };
-
-        Box::new(Swap::start_in(start_state, context))
+    pub fn proposed(request: Request<AL, BL, AA, BA>, secret_source: impl SecretSource) -> Self {
+        Self {
+            swap_communication: SwapCommunication::Proposed { request },
+            alpha_ledger_state: LedgerState::NotDeployed,
+            beta_ledger_state: LedgerState::NotDeployed,
+            secret_source: Arc::new(secret_source),
+            error: None,
+        }
     }
-
-    pub fn request(&self) -> Request<AL, BL, AA, BA> {
-        match &self.swap_communication {
-            SwapCommunication::Accepted { request, .. }
-            | SwapCommunication::Proposed { request }
-            | SwapCommunication::Declined { request, .. } => request.clone(),
+    pub fn accepted(
+        request: Request<AL, BL, AA, BA>,
+        response: AcceptResponseBody<AL, BL>,
+        secret_source: impl SecretSource,
+    ) -> Self {
+        Self {
+            swap_communication: SwapCommunication::Accepted { request, response },
+            alpha_ledger_state: LedgerState::NotDeployed,
+            beta_ledger_state: LedgerState::NotDeployed,
+            secret_source: Arc::new(secret_source),
+            error: None,
         }
     }
 
-    pub fn new(request: Request<AL, BL, AA, BA>, secret_source: Arc<dyn SecretSource>) -> Self {
+    pub fn declined(
+        request: Request<AL, BL, AA, BA>,
+        response: DeclineResponseBody,
+        secret_source: impl SecretSource,
+    ) -> Self {
         Self {
-            swap_communication: SwapCommunication::Proposed { request },
+            swap_communication: SwapCommunication::Declined { request, response },
             alpha_ledger_state: LedgerState::NotDeployed,
             beta_ledger_state: LedgerState::NotDeployed,
-            secret_source,
+            secret_source: Arc::new(secret_source),
```

Comment:
It's definitely strange that you need the secret source when instantiating the `State` as declined.

Reply:
Yeah, future refactoring I'd say :D